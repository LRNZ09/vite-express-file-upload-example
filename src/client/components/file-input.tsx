import type { PropsWithChildren } from 'react';
import type { SetRequired } from 'type-fest';

import { Input } from './ui/input';
import { Label } from './ui/label';

export type FileInputProps = PropsWithChildren<
	SetRequired<
		Pick<
			React.InputHTMLAttributes<HTMLInputElement>,
			'accept' | 'id' | 'onChange'
		>,
		'id'
	>
>;

export const FileInput = ({
	children,
	id: inputId,
	...props
}: FileInputProps) => (
	<>
		<Label htmlFor={inputId}>{children}</Label>
		<Input
			id={inputId}
			type="file"
			aria-label={children ? undefined : 'File input'}
			{...props}
		/>
	</>
);
