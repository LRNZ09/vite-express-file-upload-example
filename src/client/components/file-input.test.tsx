import { fc, test } from '@fast-check/vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect } from 'vitest';

import { FileInput } from './file-input';

const labelArb = fc.lorem({ mode: 'words' });
const inputIdAttributeArb = fc.stringMatching(/^[a-z]+(-[a-z]+)*$/);
const acceptAttributeArb = fc.string();
const fileContentArb = fc.lorem({ mode: 'sentences' });

describe('FileInput', () => {
	test('match snapshot', () => {
		const { container } = render(<FileInput id="file-input-id" />);

		expect(container).toMatchSnapshot();
	});

	test.prop([inputIdAttributeArb])('handle input id correctly', (inputId) => {
		render(<FileInput id={inputId} />);

		const inputElement = screen.getByLabelText('File input');

		expect(inputElement).toHaveAttribute('id', inputId);
	});

	test.prop([inputIdAttributeArb, labelArb], {
		examples: [
			['select-file-input-id', 'Select file'],
			['choose-file-input-id', 'Choose file'],
			['upload-file-input-id', 'Upload file'],
		],
	})('handle label correctly', (inputId, label) => {
		render(<FileInput id={inputId}>{label}</FileInput>);

		const inputElement = screen.queryByLabelText(label);

		expect(inputElement).toBeInTheDocument();
	});

	test.prop([inputIdAttributeArb, acceptAttributeArb], {
		examples: [
			['image-file-input-id', 'image/*'],
			['audio-file-input-id', 'audio/*'],
			['video-file-input-id', 'video/*'],
			['pdf-file-input-id', '.pdf'],
		],
	})('handle accept correctly', (inputId, accept) => {
		render(<FileInput id={inputId} accept={accept} />);

		const inputElement = screen.getByLabelText('File input');

		expect(inputElement).toHaveAttribute('accept', accept);
	});

	test.prop([
		inputIdAttributeArb,
		fc.func(fc.constant(undefined)),
		fileContentArb,
	])(
		'handle when a file is chosen',
		async (inputId, handleChange, fileContent) => {
			const user = userEvent.setup();

			render(<FileInput id={inputId} onChange={handleChange} />);

			const inputElement = screen.getByLabelText('File input');

			const file = new File([fileContent], `${inputId}-file.txt`, {
				type: 'text/plain',
			});
			await user.upload(inputElement, file);

			expect((inputElement as HTMLInputElement).files).toContainEqual(file);
		},
	);
});
