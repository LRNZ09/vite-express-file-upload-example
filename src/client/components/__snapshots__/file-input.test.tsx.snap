// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileInput > match snapshot 1`] = `
<div>
  <label
    class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
    for="file-input-id"
  />
  <input
    aria-label="File input"
    class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
    id="file-input-id"
    type="file"
  />
</div>
`;
